# **DERRICK GACHERU**
# **SCT211-0004/2021**
## **TITLE: Application of Newton-Raphson Method in Finding Roots of Equations**

### **INTRODUCTION**

The Newton-Raphson method is a powerful numerical technique used for finding successively better approximations to the roots of a real-valued function. Named after Sir Isaac Newton and Joseph Raphson, this iterative algorithm is widely applied in various fields of science and engineering due to its efficiency and rapid convergence rate.

In this report, we explore the principles, mathematical foundation, implementation, and applications of the Newton-Raphson method. We delve into the theoretical background of the method, discussing its formula, convergence criteria, and potential limitations. Additionally, we demonstrate the practical utility of the method through numerical examples and real-world problems.

The report aims to provide a comprehensive understanding of the Newton-Raphson method, equipping readers with the knowledge and skills necessary to apply this technique in solving nonlinear equations, optimization problems, and other numerical tasks. Through clear explanations, illustrative examples, and practical insights, readers will gain insight into the inner workings of the Newton-Raphson method and its significance in numerical analysis.

### **THEORY**

The Newton-Raphson method, also known as the Newton method, is an iterative numerical technique used to find the roots of a real-valued function. It is based on the idea of linear approximation and employs calculus principles to converge towards the roots of the function. Here, we discuss the mathematical foundation of the Newton-Raphson method, including its formula, convergence criteria, and potential limitations.

#### **Formula**:
The Newton-Raphson method iteratively refines an initial guess $(x_0)$
  to approximate a root of the function $f(x)$. The iterative formula for the $n+1$-th approximation $x_{n+1}$ is given by:
  $$ x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)} $$

where:


*   $x_{n}$   is the $n$-th approximation of the root,
*   $f(x_n)$ is the function value at $x_{n}$ and,

*   $f'(x_n)$ is the derivative of the function evaluated at $x_{n}$


Here's a brief overview of how the Newton-Raphson method works:

1. Start with an initial guess $(x_0)$ for the root of the function.
2. Compute the value of the function $f(x_0)$ and its derivative $f'(x_0)$ at the initial guess.
3. Use the formula to find a better approximation $(x_1)$ for the root:
$$ x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)} $$
4. Repeat steps 2 and 3 with the new approximation $(x_1)$ until the desired level of accuracy is achieved or until convergence criteria are met.


#### **Convergence Criteria:**
The Newton-Raphson method converges to a root of the function when the initial guess $x_0$ is sufficiently close to the actual root and the function $f(x)$ is continuous and differentiable in the neighborhood of the root. Convergence criteria include:

1. **Local Convergence:** The method converges quadratically (i.e., the number of correct digits approximately doubles with each iteration) if the initial guess is sufficiently close to the root and the function has a simple root.

2. **Global Convergence:** Convergence is guaranteed if the initial guess is in a region where the function is strictly concave or convex and the derivative does not vanish.

#### **Potential Limitations:**
While the Newton-Raphson method is a powerful tool for root-finding, it has certain limitations:

1. **Sensitivity to Initial Guess:** The method may fail to converge or converge to a different root if the initial guess is far from the actual root or lies on a critical point (where the derivative vanishes).
2. **Derivative Evaluation:** Computing the derivative of the function may be computationally expensive or numerically unstable, especially for complex functions or functions with discontinuities.
3. **Local Convergence:** The method may converge to a local minimum or maximum if the initial guess is near a critical point, rather than the desired root.

In summary, the Newton-Raphson method provides an efficient and effective means of approximating roots of real-valued functions, but careful consideration of the initial guess, function properties, and convergence criteria is essential to ensure accurate and reliable results.

### **IMPLEMENTATION IN PYTHON**

Below is the Python implementation of the Newton-Raphson method to solve a specific equation:



In [1]:
def newton_raphson(func, func_derivative, x0, tol=1e-6, max_iter=100):
    """
    Newton-Raphson method for finding a root of a function.

    Parameters:
        func : function
            The function for which to find the root.
        func_derivative : function
            The derivative of the function.
        x0 : float
            Initial guess for the root.
        tol : float, optional
            Tolerance (stopping criterion) for the root. Default is 1e-6.
        max_iter : int, optional
            Maximum number of iterations. Default is 100.

    Returns:
        x : float
            Approximation of the root.
        num_iter : int
            Number of iterations performed.
    """
    x = x0
    num_iter = 0
    while abs(func(x)) > tol and num_iter < max_iter:
        x -= func(x) / func_derivative(x)
        num_iter += 1
    return x, num_iter

# Example usage:
def f(x):
    return x**2 - 2

def f_prime(x):
    return 2*x

root, num_iterations = newton_raphson(f, f_prime, 1.5)
print("Root:", root)
print("Number of iterations:", num_iterations)


Root: 1.4142135623746899
Number of iterations: 3


### ***Explanation***

1. **Function Definition**:
    - The `newton_raphson` function takes four arguments: `func`, the function for which to find the root; `func_derivative`, the derivative of the function; `x0`, the initial guess for the root; `tol`, the tolerance for the root (defaults to $10^{-6}$); `max_iter`, the maximum number of iterations (defaults to 100).
2. **Newton-Raphson Method**:
    - The function iteratively applies the Newton-Raphson formula $$ x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)} $$ until either the absolute value of the function value is less than the tolerance (`tol`) or the maximum number of iterations (`max_iter`) is reached.
3. **Example Usage**:
    - An example usage is provided for finding the square root of 2 using the `f(x) = x^2 - 2` function and its derivative `f'(x) = 2x`. The root and the number of iterations are printed.

This implementation allows for finding roots of various functions using the Newton-Raphson method.
$10^{-6}$




### **RESULTS AND ANALYSIS**

The output of the Python program was:

- **Root:** $1.4142135623746899$
- **Number of iterations:** $3$

**Efficiency:** The method demonstrated efficiency by converging to the root in a small number of iterations. This suggests that the Newton-Raphson method can quickly find solutions to equations, making it suitable for real-time or iterative calculations.

**Accuracy:** The accuracy of the solution depends on factors such as the initial guess, the behavior of the function near the root, and the numerical precision of the calculations. In this case, the obtained root appears to be accurate, but further analysis may be needed to assess its precision under different conditions.

**Convergence:** The method exhibited convergence, as evidenced by the decreasing number of iterations required to reach the root. Convergence indicates that the method is approaching the true solution with each iteration, which is a desirable property for numerical algorithms.

**Challenges:** While the Newton-Raphson method is generally robust and efficient, challenges may arise in certain scenarios. For example, the method may fail to converge if the initial guess is far from the true root or if the function has complex behavior near the root, such as oscillations or singularities. Additionally, issues related to numerical stability or precision may affect the accuracy of the solution.

Overall, the Newton-Raphson method appears to be effective in this scenario, but further testing and analysis may be necessary to assess its performance comprehensively.

### **CONCLUSION**


In conclusion, the Newton-Raphson method proved to be an efficient and effective numerical technique for finding roots of equations. The method demonstrated rapid convergence and accuracy in finding the root of the given equation. Despite its simplicity, the Newton-Raphson method's quadratic convergence rate makes it particularly useful for problems where fast convergence is desired.

The applicability of the Newton-Raphson method extends across various fields, including scientific computing, engineering, physics, finance, and optimization. It is commonly used in solving nonlinear equations, optimization problems, and numerical analysis tasks. The method's simplicity, efficiency, and robust convergence properties make it a valuable tool in both theoretical and practical contexts.

However, it's important to note that the Newton-Raphson method may encounter challenges such as divergence, sensitivity to initial guesses, and difficulties with functions exhibiting complex behavior or singularities. Careful consideration of these factors and appropriate adjustments to the implementation can help mitigate potential issues and enhance the method's applicability.

Overall, the Newton-Raphson method remains a fundamental and widely used technique in numerical analysis, offering a balance of efficiency, accuracy, and simplicity that makes it invaluable for solving a variety of mathematical problems across different disciplines.

### REFERENCES



1. Burden, R. L., & Faires, J. D. (2010). *Numerical Analysis (9th ed.).* Brooks/Cole.
2. Quarteroni, A., Sacco, R., & Saleri, F. (2007). *Numerical Mathematics (2nd ed.).* Springer.
3. Stoer, J., & Bulirsch, R. (2002). *Introduction to Numerical Analysis (3rd ed.).* Springer.

These references were consulted for theoretical concepts, numerical methods, and implementation techniques related to the Newton-Raphson method and numerical analysis in general.
